firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/FirebaseCrashlyticsReportManager.java

```diff
@@ -0,0 +1,102 @@
+// limitations under the License.
+
+package com.google.firebase.crashlytics.internal.common;
+
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.crashlytics.internal.model.CrashlyticsReport;
+import com.google.firebase.crashlytics.internal.persistence.CrashlyticsReportPersistence;
+import com.google.firebase.crashlytics.internal.send.DataTransportCrashlyticsReportSender;
+import com.google.firebase.crashlytics.internal.settings.model.AppSettingsData;
+import java.util.List;
+
+/**
+ * This class handles Crashlytics lifecycle events and manages capture, persistence, and sending of
+ * reports to Firebase Crashlytics.
+ */
+public class FirebaseCrashlyticsReportManager {
+
+  private static final String EVENT_TYPE_CRASH = "crash";
+  private static final int EVENT_THREAD_IMPORTANCE = 4;
+  private static final int MAX_CHAINED_EXCEPTION_DEPTH = 8;
+
+  private final CrashlyticsReportDataCapture dataCapture;
+  private final CrashlyticsReportPersistence reportPersistence;
+  private final DataTransportCrashlyticsReportSender reportsSender;
+  private final CurrentTimeProvider currentTimeProvider;
+
+  private String currentSessionId;
+
+  public FirebaseCrashlyticsReportManager(
+      CrashlyticsReportDataCapture dataCapture,
+      CrashlyticsReportPersistence reportPersistence,
+      DataTransportCrashlyticsReportSender reportsSender,
+      CurrentTimeProvider currentTimeProvider) {
+    this.dataCapture = dataCapture;
+    this.reportPersistence = reportPersistence;
+    this.reportsSender = reportsSender;
+    this.currentTimeProvider = currentTimeProvider;
+  }
+
+  public void onSessionBegin(String sessionId) {
+    final long timestamp = currentTimeProvider.getCurrentTimeMillis() / 1000;
+    currentSessionId = sessionId;
+
+    final CrashlyticsReport capturedReport = dataCapture.captureReportData(sessionId, timestamp);
+
+    reportPersistence.persistReport(capturedReport);
+  }
+
+  public void onFatalEvent(Throwable event, Thread thread) {
+    final long timestamp = currentTimeProvider.getCurrentTimeMillis() / 1000;
+
+    final CrashlyticsReport.Session.Event capturedEvent =
+        dataCapture.captureEventData(
+            event,
+            thread,
+            EVENT_TYPE_CRASH,
+            timestamp,
+            EVENT_THREAD_IMPORTANCE,
+            MAX_CHAINED_EXCEPTION_DEPTH);
+
+    reportPersistence.persistEvent(capturedEvent, currentSessionId);
+  }
+
+  public void onSessionEnd() {
+    currentSessionId = null;
+  }
+
+  public void onSessionsFinalize() {
+    reportPersistence.finalizeReports(currentSessionId);
+  }
+
+  public void onReportSend(AppSettingsData appSettingsData) {
+    final List<CrashlyticsReport> reportsToSend = reportPersistence.loadFinalizedReports();
+    for (CrashlyticsReport report : reportsToSend) {
+      reportsSender
+          .sendReport(report.withOrganizationId(appSettingsData.organizationId))
+          .addOnCompleteListener(this::onReportSendComplete);
```

Comment:
It's usually kind of a smell to use `addOnCompleteListener` instead of `continueWith`, but it's not obviously wrong here. The implications are:
1. All of the sends will run in parallel.
2. This caller won't ever know if they succeeded or failed (or even finished).
3. The callback will happen on the UI Thread.

Implication #2 seems intentional, and #3 seems unintentional. I'm not sure about #1. It might be a good idea to add a comment about the threading expectations of this method.

Reply:
Hmm, I was just following the instructions here: https://developers.google.com/android/guides/tasks

Definitely good points. I'll follow up with you on this.